protobuf/src/stream.rs

```diff
@@ -703,147 +697,22 @@ impl<'a> WithCodedInputStream for &'a Bytes {
     }
 }
 
-enum OutputTarget<'a> {
-    Write(&'a mut Write, Vec<u8>),
-    Vec(&'a mut Vec<u8>),
-    Bytes,
-}
-
 pub struct CodedOutputStream<'a> {
-    target: OutputTarget<'a>,
-    // alias to buf from target
-    buffer: &'a mut [u8],
-    // within buffer
-    position: usize,
+    target: BufWriter<&'a mut Write>,
```

Comment:
As a slight optimisation here we could add a `T: Write` type parameter to `CodedOutputStream` and then replace this with `BufWriter<&'a mut T>`. This would allow the compiler to inline writes to vecs and slices.

Reply:
However it looks like this would affect generated code, so it may not be a feasible option.